File: framework/data-jpa/src/main/java/top/bettercode/simpleframework/data/jpa/IBaseService.java
package top.bettercode.simpleframework.data.jpa;

import java.util.List;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.lang.Nullable;

/**
 * @author Peter Wu
 */
public interface IBaseService<T, ID, M extends BaseRepository<T, ID>> {

  M getRepository();

  <S extends T> S save(S s);

  <S extends T> int save(S s, Specification<T> spec);

  /**
   * 动态更新，只更新非Null字段
   *
   * @param s   对象
   * @param <S> 类型
   * @return 结果
   * @deprecated 不建议再使用, 请使用以下方式替代
   * <p>
   * entity.nullPropertySetFrom(exist);
   * <p>
   * save(entity);
   * <p>
   */
  @Deprecated
  <S extends T> S dynamicSave(S s);

  <S extends T> List<S> saveAll(Iterable<S> entities);

  void delete(T t);

  void deleteById(ID id);

  int delete(Specification<T> spec);

  void deleteAllById(Iterable<? extends ID> ids);

  void deleteAll(Iterable<? extends T> entities);

  void deleteAll();

  void deleteAllInBatch(Iterable<T> entities);

  void deleteAllInBatch();

  long count();

  Optional<T> findById(ID id);

  Optional<T> findFirst(Sort sort);

  boolean existsById(ID id);

  List<T> findAll();

  List<T> findAllById(Iterable<ID> ids);

  List<T> findAll(int size);

  List<T> findAll(int size, Sort sort);

  Page<T> findAll(Pageable pageable);

  List<T> findAll(Sort sort);

  long count(@Nullable Specification<T> spec);

  boolean exists(Specification<T> spec);

  Optional<T> findOne(@Nullable Specification<T> spec);

  Optional<T> findFirst(Specification<T> spec);

  List<T> findAll(@Nullable Specification<T> spec);

  List<T> findAll(Specification<T> spec, int size);

  List<T> findAll(Specification<T> spec, int size, Sort sort);

  Page<T> findAll(@Nullable Specification<T> spec, Pageable pageable);

  List<T> findAll(@Nullable Specification<T> spec, Sort sort);

}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/IBaseService.kt
package top.bettercode.summer.data.jpa

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.domain.Specification
import org.springframework.transaction.annotation.Transactional
import top.bettercode.summer.data.jpa.support.UpdateSpecification
import java.util.*

/**
 * @author Peter Wu
 */
interface IBaseService<T:Any, ID:Any, M : BaseRepository<T, ID>> {

    fun getRepository(): M

    fun <E> findAllPageByPage(totalPages: Int, query: (Pageable) -> List<E>): List<E>

    fun <E> findAllPageByPage(pageSize: Int, totalPages: Int, query: (Pageable) -> List<E>): List<E>

    fun <S : T> save(s: S): S

    /**
     * 动态更新，只更新非Null字段
     *
     * @param s   对象
     * @param <S> 类型
     * @return 结果
    </S> */
    fun <S : T> saveDynamic(s: S): S

    fun <S : T> saveAll(entities: Iterable<S>): List<S>

    @Transactional
    fun update(spec: UpdateSpecification<T>): Long

    @Transactional
    fun <S : T> update(s: S?, spec: UpdateSpecification<T>): Long

    fun delete(t: T)
    fun deleteById(id: ID)
    fun delete(spec: Specification<T>): Long
    fun deleteAllById(ids: Iterable<ID>)
    fun deleteAll(entities: Iterable<T>)
    fun deleteAll()
    fun deleteAllInBatch(entities: Iterable<T>)
    fun deleteAllInBatch()
    fun count(): Long
    fun findById(id: ID): Optional<T>
    fun findFirst(sort: Sort): Optional<T>
    fun existsById(id: ID): Boolean
    fun findAll(): List<T>
    fun findAllById(ids: Iterable<ID>): List<T>

    fun findAll(size: Int): List<T>
    fun findAll(offset: Long, size: Int): List<T>
    fun findAll(size: Int, sort: Sort): List<T>
    fun findAll(offset: Long, size: Int, sort: Sort): List<T>

    fun findAll(pageable: Pageable): Page<T>
    fun findAll(sort: Sort): List<T>
    fun count(spec: Specification<T>?): Long
    fun exists(spec: Specification<T>): Boolean
    fun findOne(spec: Specification<T>?): Optional<T>
    fun findFirst(spec: Specification<T>?): Optional<T>

    fun findAll(spec: Specification<T>?): List<T>

    fun findAll(spec: Specification<T>?, size: Int): List<T>
    fun findAll(spec: Specification<T>?, offset: Long, size: Int): List<T>
    fun findAll(spec: Specification<T>?, size: Int, sort: Sort): List<T>
    fun findAll(spec: Specification<T>?, offset: Long, size: Int, sort: Sort): List<T>

    fun findAll(spec: Specification<T>?, pageable: Pageable): Page<T>
    fun findAll(spec: Specification<T>?, sort: Sort): List<T>
}
